### ISSUE TITLE
Run trivially_copy_pass_by_ref for traits?

### ISSUE THREAD

**Issue by crumblingstatue** [rust-lang/rust-clippy#3318]
```Rust
#[derive(Copy, Clone)]
pub struct Color {
    pub r: u8, pub g: u8, pub b: u8, pub a: u8,
}

pub struct Foo;

impl Foo {
    pub fn foo_method(&self, _color: &Color) {
        unimplemented!()
    }
}

pub trait MyTrait {
    fn trait_method(&self, _color: &Color);
}

fn main() {
    
}
```

Clippy lints `foo_method` for `trivially_copy_pass_by_ref`, but not `trait_method`. Is it intentionally not implemented for traits due to potential problems, or just due to an oversight?

I'm asking because it made me convert half my public API to take `Color` by value, but the other half still takes `&Color`, because those are provided by traits. This makes the API inconsistent.

**Comment by flip1995** [comment]
This is indeed intentional:
https://github.com/rust-lang-nursery/rust-clippy/blob/0f4b13bc1bbd43ed21878c21291dddde7e4b9028/clippy_lints/src/trivially_copy_pass_by_ref.rs#L120-L127

This code excludes functions/methods from `impl Trait for Foo { .. }` and `trait Trait { .. }`.

Here's the issue to the lint, where this was more or less "discussed": #1680

**Comment by crumblingstatue** [comment]
The only mention of traits is in
> Propably ignore trait or generic functions and let compiler optimisation handle this.

It doesn't mention any specific problems, it just suggests that we rely on compiler optimization.

My only problem with this is that it can make some APIs inconsistent if applied, and the maintainer has to manually review the entire API to check for inconsistencies. Perhaps there could be a `trivially_copy_pass_by_ref_traits` lint that's `allow` by default, and could assist the developer in this case.

**Comment by crumblingstatue** [comment]
> The problem with `impl Trait` is that the implemented `Trait` could be from an extern crate, where you have no influence on the definition.

Yes, to clarify, I don't mean to apply this lint for trait `impl`s, only for trait definitions.

**Comment by waynr** [comment]
Just fyi, I'm gonna try taking this on as a first issue.

**Comment by waynr** [comment]
So, getting  somewhat slow start on this (spent way too long trying to get vim to go to definition using RLS before giving up and looking up definitions online). Anyway, here is where I am at so far.

Based on the discussion so far in this thread I thought that I would have to modify the block with the `// Exclude non-inherent impls` to remove the `ItemKind::Trait(..)` pattern so that the lint would start warning on trait definitions that use a reference where it would be cheaper to copy. However, this didn't cause the test case to start failing after I updated it with an example similar to @crumblingstatue's.

So the next thing I decided to do was to try running clippy against the original example verbatim with the `ItemKind::Trait(..)` pattern removed from the `matches!` call. This also did not result in the expected warning message:

```
zsh/3 7940 [148]  (git)-[master]-% env CLIPPY_TESTS=true cargo run --bin clippy-driver -- -L ./target/debug whatever.rs
   Compiling clippy_lints v0.0.212 (/home/wayne/projects/rust-lang-nursery/rust-clippy/clippy_lints)
   Compiling clippy v0.0.212 (/home/wayne/projects/rust-lang-nursery/rust-clippy)
    Finished dev [unoptimized + debuginfo] target(s) in 42.26s             
     Running `target/debug/clippy-driver -L ./target/debug whatever.rs`   
warning: this argument is passed by reference, but would be more efficient if passed by value   
 --> whatever.rs:9:38                                                                             
  |                                                                
9 |     pub fn foo_method(&self, _color: &Color) {                                                                                    
  |                                      ^^^^^^ help: consider passing by value instead: `Color`
  |                                                                        
  = note: #[warn(clippy::trivially_copy_pass_by_ref)] on by default       
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#trivially_copy_pass_by_ref
```

Here is my `git diff` output:
```
zsh/3 7951  (git)-[master]-% git diff
diff --git a/clippy_lints/src/trivially_copy_pass_by_ref.rs b/clippy_lints/src/trivially_copy_pass_by_ref.rs
index 61a2a9de..30885c8f 100644
--- a/clippy_lints/src/trivially_copy_pass_by_ref.rs
+++ b/clippy_lints/src/trivially_copy_pass_by_ref.rs
@@ -119,8 +119,7 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for TriviallyCopyPassByRef {

         // Exclude non-inherent impls
         if let Some(Node::Item(item)) = cx.tcx.hir.find(cx.tcx.hir.get_parent_node(node_id)) {
-            if matches!(item.node, ItemKind::Impl(_, _, _, _, Some(_), _, _) |
-                ItemKind::Trait(..))
+            if matches!(item.node, ItemKind::Impl(_, _, _, _, Some(_), _, _))
             {
                 return;
             }
```

Am I doing something incorrectly here? I'll probably start looking for other early returns in the lint code to see if trait definitions could be caught up elsewhere.

**Comment by waynr** [comment]
Okay, blew a bunch more time trying to get `rust-lldb` working with the manual test command line but didn't get it. I gave up and just started using printlns and here is what I found:

The diff:
```
zsh/3 8043 [148]  (git)-[master]-% git diff                                                                            
diff --git a/clippy_lints/src/trivially_copy_pass_by_ref.rs b/clippy_lints/src/trivially_copy_pass_by_ref.rs
index 61a2a9de..d249059e 100644
--- a/clippy_lints/src/trivially_copy_pass_by_ref.rs
+++ b/clippy_lints/src/trivially_copy_pass_by_ref.rs
@@ -98,6 +98,9 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for TriviallyCopyPassByRef {
         span: Span,
         node_id: NodeId,
     ) {
+        println!("{:?}", span);
+        println!("{:?}", decl);
+
         if in_macro(span) {
             return;
         }
```

The modified example file:
```rust
zsh/3 8017  (git)-[master]-% cat whatever.rs 
#[derive(Copy, Clone)]
pub struct Color {
    pub r: u8, pub g: u8, pub b: u8, pub a: u8,
}

pub struct Foo;

impl Foo {
    pub fn foo_method(&self, _color: &Color) {
        unimplemented!()
    }
    pub fn foo_method2(&self, _color: &Color) {
        unimplemented!()
    }
}

trait MyTrait {
    fn trait_method(&self, _color: &Color);
}

pub trait MyTrait2 {
    fn trait_method2(&self, _color: &Color);
}

impl MyTrait for Foo {
    fn trait_method(&self, _color: &Color) {
        unimplemented!()
    }
}

fn main() {

}
```

The lint output:

```
zsh/3 8018  (git)-[master]-% env CLIPPY_TESTS=true cargo run --bin clippy-driver -- -L ./target/debug whatever.rs
    Finished dev [unoptimized + debuginfo] target(s) in 0.20s
     Running `target/debug/clippy-driver -L ./target/debug whatever.rs`
whatever.rs:1:16: 1:21
FnDecl { inputs: [type(&Self)], output: Return(type(Color)), variadic: false, implicit_self: ImmRef }
whatever.rs:9:5: 11:6
FnDecl { inputs: [type(&Self), type(&Color)], output: DefaultReturn(whatever.rs:9:46: 9:46), variadic: false, implicit_self: ImmRef }
warning: this argument is passed by reference, but would be more efficient if passed by value
 --> whatever.rs:9:38
  |
9 |     pub fn foo_method(&self, _color: &Color) {
  |                                      ^^^^^^ help: consider passing by value instead: `Color`
  |
  = note: #[warn(clippy::trivially_copy_pass_by_ref)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#trivially_copy_pass_by_ref

whatever.rs:12:5: 14:6
FnDecl { inputs: [type(&Self), type(&Color)], output: DefaultReturn(whatever.rs:12:47: 12:47), variadic: false, implicit_self: ImmRef }
warning: this argument is passed by reference, but would be more efficient if passed by value
  --> whatever.rs:12:39
   |
12 |     pub fn foo_method2(&self, _color: &Color) {
   |                                       ^^^^^^ help: consider passing by value instead: `Color`
   |
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#trivially_copy_pass_by_ref

whatever.rs:26:5: 28:6
FnDecl { inputs: [type(&Self), type(&Color)], output: DefaultReturn(whatever.rs:26:44: 26:44), variadic: false, implicit_self: ImmRef }
whatever.rs:31:1: 33:2
FnDecl { inputs: [], output: DefaultReturn(whatever.rs:31:11: 31:11), variadic: false, implicit_self: None }
```

What's demonstrated here if you look carefully at the line numbers in the println outputs and the declarations is that the trait definitions themselves are missing.

This makes me think that the `check_fn` method never sees function declarations found in trait blocks, only those found in impl blocks. I'm about to step away from the computer for a while, but my next step in figuring this out will be to figure out which `LateLintPass` functions do visit traits and figure out how to incorporate one of those into this lint.

**Comment by waynr** [comment]
@flip1995 is there anything left for me to do for this issue?

**Comment by flip1995** [comment]
Nope this is resolved by #3437. Thanks for reminding me!
